Update vector db quickstart guide (#374)
redis/docs: content/develop/get-started/vector-database.md

This quick start guide helps you to:

1. Understand what a vector database is
2. Create a Redis vector database
3. Create vector embeddings and store vectors
4. Query data and perform a vector search

## Understand vector databases

> Data is often unstructured, which means that it isn't described by a well-defined schema. Examples of unstructured data include text passages, images, videos, or audio. An approach to storing and searching through unstructured data is through vector embeddings.

Reviewer: ```suggestion
Data is often unstructured, which means that it isn't described by a well-defined schema. Examples of unstructured data include text passages, images, videos, or audio. One approach to storing and searching through unstructured data is to use vector embeddings.
```

**What are vectors?** In machine learning and AI, vectors are sequences of numbers that represent data. They are the inputs and outputs of models, encapsulating underlying information in a numerical form. Vectors transform unstructured data, such as text, images, videos, and audio, into a format that machine learning models can process.

> - **Why are They Important?** Vectors capture complex patterns and semantic meanings inherent in data, making them powerful tools for a variety of applications. They allow machine learning models to understand and manipulate unstructured data more effectively.

Reviewer: ```suggestion
- **Why are they important?** Vectors capture complex patterns and semantic meanings inherent in data, making them powerful tools for a variety of applications. They allow machine learning models to understand and manipulate unstructured data more effectively.
```

> - **Enhancing Traditional Search?** Traditional keyword or lexical search relies on exact matches of words or phrases, which can be limiting. In contrast, vector search, or semantic search, leverages the rich information captured in vector embeddings. By mapping data into a vector space, similar items are positioned near each other based on their meaning. This approach allows for more accurate and meaningful search results, as it considers the context and semantic content of the query rather than just the exact words used.

Reviewer: ```suggestion
- **Enhancing traditional search.** Traditional keyword or lexical search relies on exact matches of words or phrases, which can be limiting. In contrast, vector search, or semantic search, leverages the rich information captured in vector embeddings. By mapping data into a vector space, similar items are positioned near each other based on their meaning. This approach allows for more accurate and meaningful search results, as it considers the context and semantic content of the query rather than just the exact words used.
```

    "material": "aluminium",
    "weight": "10"
  },
  "description": "Small and powerful, the Jigger is the best ride for the smallest of tikes! ..."
}
```

> The `description` field contains a free-form text descriptions of bikes and will be used to create vector embeddings.

Reviewer: ```suggestion
The `description` field contains free-form text descriptions of bikes and will be used to create vector embeddings.
```

###  1. Fetch the demo data
You need to first fetch the demo dataset as a JSON array:

{{< clients-example search_vss get_data />}}

Inspect the structure of one of the bike JSON documents:

{{< clients-example search_vss dump_data />}}

### 2. Store the demo data in Redis
Now iterate over the `bikes`  array to store the data as [JSON]({{< relref "/develop/data-types/json/" >}}) documents in Redis by using the [JSON.SET]({{< baseurl >}}/commands/json.set//) command. The below code uses a [pipeline]({{< relref "/develop/use/pipelining" >}}) to minimize the network round-trip times:

{{< clients-example search_vss load_data />}}

Once loaded, you can retrieve a specific attributes from one of the JSON documents in Redis using a [JSONPath](https://goessner.net/articles/JsonPath/) expression:

{{< clients-example search_vss get />}}

### 3. Select a text embedding model

> [HuggingFace](https://huggingface.co) has a large catalog of text embedding models that are locally servable through the `SentenceTransformers` framework. Here we use the [MS MARCO](https://microsoft.github.io/msmarco/) model which is widely used in search engines, chatbots, and other AI applications.

Reviewer: ```suggestion
[HuggingFace](https://huggingface.co) has a large catalog of text embedding models that are locally servable through the `SentenceTransformers` framework. Here we use the [MS MARCO](https://microsoft.github.io/msmarco/) model that is widely used in search engines, chatbots, and other AI applications.
```

```python
from sentence_transformers import SentenceTransformer

embedder = SentenceTransformer('msmarco-distilbert-base-v4')
```

### 4. Generate text embeddings
Iterate over all the Redis keys with the prefix `bikes:`:

{{< clients-example search_vss get_keys />}}

Use the keys as input to the [JSON.MGET]({{< baseurl >}}/commands/json.mget//) command, along with the `$.description` field, to collect the descriptions in a list. Then, pass the list of descriptions to the `.encode()` method:

{{< clients-example search_vss generate_embeddings />}}

> Insert the vectorized descriptions to the bike documents in Redis using the [JSON.SET]({{< baseurl >}}/commands/json.set//) command. The following command inserts a new field in each of the documents under the JSONPath `$.description_embeddings`. Once again, do this using a pipeline to avoid unnecessary network round-trips:

Reviewer: ```suggestion
Insert the vectorized descriptions to the bike documents in Redis using the [JSON.SET]({{< baseurl >}}/commands/json.set) command. The following command inserts a new field into each of the documents under the JSONPath `$.description_embeddings`. Once again, do this using a pipeline to avoid unnecessary network round-trips:
```

{{< clients-example search_vss load_embeddings />}}

> Inspect one of the updated bike documents using the [JSON.GET](({{< relref "/commands" >}}/json.get/)) command:

Reviewer: ```suggestion
Inspect one of the updated bike documents using the [JSON.GET]({{< baseurl >}}/commands/json.get) command:
```

{{< clients-example search_vss dump_example />}}

{{% alert title="Note" color="warning" %}}
When storing a vector embedding within a JSON document, the embedding is stored as a JSON array. In the example above, the array was shortened considerably for the sake of readability.
{{% /alert  %}}

## Create an index

### 1. Create an index with a vector field

> You must create an index to query based on document metadata or perform vector searches. Use the [FT.CREATE]({{< baseurl >}}/commands/ft.create//) command:

Reviewer: ```suggestion
You must create an index to query document metadata or to perform vector searches. Use the [FT.CREATE]({{< baseurl >}}/commands/ft.create) command:
```

  SCHEMA
    $.model TEXT WEIGHT 1.0 NOSTEM
    $.brand TEXT WEIGHT 1.0 NOSTEM
    $.price NUMERIC
    $.type TAG SEPARATOR ","
    $.description AS description TEXT WEIGHT 1.0
    $.description_embeddings AS vector VECTOR FLAT 6 TYPE FLOAT32 DIM 768 DISTANCE_METRIC COSINE
{{< /clients-example >}}

Here is a breakdown of the `VECTOR` field definition:

* `$.description_embeddings AS vector`: The vector field's JSON path and its field alias `vector`.
* `FLAT`: Specifies the indexing method, which is either a flat index or a hierarchical navigable small world graph ([HNSW](https://arxiv.org/ftp/arxiv/papers/1603/1603.09320.pdf)).
* `TYPE FLOAT32`: Sets the float precision of a vector component, in this case a 32-bit floating point number.
* `DIM 768`: The length or dimension of the embeddings, determined by the chosen embedding model.
* `DISTANCE_METRIC COSINE`: The chosen distance function: [cosine distance](https://en.wikipedia.org/wiki/Cosine_similarity).

You can find further details about all these options in the [vector reference documentation]({{< relref "/develop/interact/search-and-query/advanced-concepts/vectors" >}}).

### 2. Check the state of the index

> As soon as you execute the [FT.CREATE]({{< baseurl >}}/commands/ft.create//) command, the indexing process runs in the background. In a short time, all JSON documents should be indexed and ready to be queried. To validate that, you can use the [FT.INFO]({{< baseurl >}}/commands/ft.info//) command, which provides details and statistics about the index. Of particular interest are the number of documents successfully indexed and the number of failures:

Reviewer: ```suggestion
As soon as you execute the [FT.CREATE]({{< baseurl >}}/commands/ft.create) command, the indexing process runs in the background. In a short time, all JSON documents should be indexed and ready to be queried. To validate that, you can use the [FT.INFO]({{< baseurl >}}/commands/ft.info) command, which provides details and statistics about the index. Of particular interest are the number of documents successfully indexed and the number of failures:
```

{{< clients-example search_vss validate_index >}}
FT.INFO idx:bikes_vss
{{< /clients-example >}}

## Perform vector searches

> This quick start guide focuses on vector search. However, you can learn more about how to query based on vector metadata in the [document database quick start guide]({{< relref "/develop/get-started/document-database" >}}).

Reviewer: Did you mean "... you can learn more about how to query document metadata in the ..." ?

Author: Yep!

> This quick start guide focuses on vector search. However, you can learn more about how to query based on vector metadata in the [document database quick start guide]({{< relref "/develop/get-started/document-database" >}}).

Author: ```suggestion
This quick start guide focuses on vector search. However, you can learn more about how to query based on document metadata in the [document database quick start guide]({{< relref "/develop/get-started/document-database" >}}).
```
```suggestion
This quick start guide focuses on vector search. However, you can learn more about how to query based on vector metadata in the [document database quick start guide]({{< relref "/develop/get-started/document-database" >}}).
```

)
```

Let's break down the above query template:

> - The filter expression `(*)` means `all`. In other words, NO filtering was applied. You could replace it with an expression that filters by additional metadata.

Reviewer: ```suggestion
- The filter expression `(*)` means `all`. In other words, no filtering was applied. You could replace it with an expression that filters by additional metadata.
```
